Core/src/core/smallEvents/space.ts

```diff
@@ -0,0 +1,129 @@
+	if (neoWSFeed.length > 0) {
+		const randomObject: NearEarthObject = RandomUtils.draftbotRandom.pick(neoWSFeed);
+		return {
+			mainValue: neoWSFeed.length,
+			randomObjectName: randomObject.name,
+			randomObjectDistance: Math.floor(parseInt(randomObject.close_approach_data[0].miss_distance.kilometers, 10) / 1000000),
+			randomObjectDiameter: Math.floor((randomObject.estimated_diameter.meters.estimated_diameter_max + randomObject.estimated_diameter.meters.estimated_diameter_min) / 2)
+		};
+	}
+	// If the list is empty, return a random invented object
+	return {
+		mainValue: 1,
+		randomObjectName: RandomUtils.draftbotRandom.pick(SpaceConstants.INVENTED_ASTEROIDS_NAMES),
+		randomObjectDistance: RandomUtils.rangedInt(SpaceConstants.DISTANCE_RANGE),
+		randomObjectDiameter: RandomUtils.rangedInt(SpaceConstants.DIAMETER_RANGE)
+	};
+}
+
+/**
+ * Gives when the next moon phase is occurring
+ */
+function moonPhase(): SpaceFunctionResult {
+	return {mainValue: SearchMoonQuarter(new Date()).quarter};
+}
+
+/**
+ * Gives when the next full moon is occurring
+ */
+function nextFullMoon(): SpaceFunctionResult {
+	let days = 0;
+	const currDate = new Date();
+	let currDegrees = MoonPhase(currDate);
+	let nextDegrees = MoonPhase(new Date(currDate.getDate() + 1));
+	while (currDegrees > 180 || nextDegrees <= 180) {
+		currDegrees = nextDegrees;
+		currDate.setDate(currDate.getDate() + 1);
+		nextDegrees = MoonPhase(currDate);
+		days++;
+	}
+	return {mainValue: days};
+}
+
+/**
+ * Gives when the next eclipse of a kind is occurring
+ */
+function nextEclipse(kind: string): SpaceFunctionResult {
+	let eclipse = SearchLunarEclipse(new Date());
+	while (eclipse.kind !== kind) {
+		eclipse = NextLunarEclipse(eclipse.peak);
+	}
+	return {mainValue: Math.floor((eclipse.peak.date.valueOf() - new Date().valueOf()) / (1000 * 3600 * 24))};
+}
+
+/**
+ * Gives when the next partial lunar eclipse is coming
+ */
+function nextPartialLunarEclipse(): SpaceFunctionResult {
+	return nextEclipse(SpaceConstants.PARTIAL_LUNAR_ECLIPSE);
+}
+
+/**
+ * Gives when the next total lunar eclipse is coming
+ */
+function nextTotalLunarEclipse(): SpaceFunctionResult {
+	return nextEclipse(SpaceConstants.TOTAL_LUNAR_ECLIPSE);
+}
+
+// Map the functions to the keys
+const spaceFunctions = {
+	neoWS,
+	moonPhase,
+	nextFullMoon,
+	nextPartialLunarEclipse,
+	nextTotalLunarEclipse
+};
+
+export const smallEventFuncs: SmallEventFuncs = {
+	canBeExecuted: Maps.isOnContinent,
+
+	executeSmallEvent(context, response): void {
+		response.push(makePacket(SmallEventSpaceInitialPacket, {}));
+		new Promise(() => {
+			let availableSpaceFunctions = Object.keys(spaceFunctions);
+			if (nextFullMoon().mainValue === 0) {
+				availableSpaceFunctions = availableSpaceFunctions.filter(e => e !== SpaceConstants.FUNCTIONS.nextFullMoon);
+			}
+
+			const specificEvent = RandomUtils.draftbotRandom.pick(availableSpaceFunctions) as keyof typeof spaceFunctions;
+			const t0 = performance.now();
```

Comment:
c'est quoi ce truc performance.now ?

Reply:
> c'est quoi ce truc performance.now ?

ca date de l'ancien mini event, c'est une sorte de Date.now() en plus précis, j'ai pas osé y toucher

Reply:
mdr ok if it works